sdks/python/apache_beam/transforms/environments.py

```diff
@@ -0,0 +1,258 @@
+    return cls(container_image=options.environment_config)
+
+@Environment.register_environment
+class ProcessEnvironment(Environment):
+
+  urn = common_urns.environments.PROCESS.urn
+
+  def __init__(self, command, os='', arch='', env=None):
+    self.command = command
+    self.os = os
+    self.arch = arch
+    self.env = env or {}
+
+  def __eq__(self, other):
+    return self.__class__ == other.__class__ \
+      and self.command == other.command and self.os == other.os \
+      and self.arch == other.arch and self.env == other.env
+
+  def to_runner_api_parameter(self, context):
+    return (common_urns.environments.PROCESS.urn,
+            beam_runner_api_pb2.ProcessPayload(
+                os=self.os,
+                arch=self.arch,
+                command=self.command,
+                env=self.env
+            ))
+
+  @Environment.register_urn(common_urns.environments.PROCESS.urn,
+                            beam_runner_api_pb2.ProcessPayload)
+  def from_runner_api_parameter(payload, components, context):
+    return ProcessEnvironment(command=payload.command, os=payload.os,
+                              arch=payload.arch, env=payload.env)
+
+  @classmethod
+  def from_options(cls, options):
+    config = json.loads(options.environment_config)
+    return cls(config.get('command'), os=config.get('os', ''),
+               arch=config.get('arch', ''), env=config.get('env', ''))
+
+@Environment.register_environment
+class ExternalEnvironment(Environment):
+
+  urn = common_urns.environments.EXTERNAL.urn
+
+  def __init__(self, url, params=None):
+    self.url = url
+    self.params = params
+
+  def __eq__(self, other):
+    return self.__class__ == other.__class__ and self.url == other.url \
+      and self.params == other.params
+
+  def to_runner_api_parameter(self, context):
+    return (common_urns.environments.EXTERNAL.urn,
+            beam_runner_api_pb2.ExternalPayload(
+                endpoint=endpoints_pb2.ApiServiceDescriptor(url=self.url),
+                params=self.params
+            ))
+
+  @Environment.register_urn(common_urns.environments.EXTERNAL.urn,
+                            beam_runner_api_pb2.ExternalPayload)
+  def from_runner_api_parameter(payload, components, context):
+    return ExternalEnvironment(payload.endpoint.url, params=payload.params)
+
+  @classmethod
+  def from_options(cls, options):
+    def looks_like_json(environment_config):
+      import re
+      return re.match(r'\s*\{.*\}\s*$', environment_config)
+
+    if looks_like_json(options.environment_config):
+      config = json.loads(options.environment_config)
+      url = config.get('url')
+      if not url:
+        raise ValueError('External environment endpoint must be set.')
+      params = config.get('params')
+    else:
+      url = options.environment_config
+      params = None
+
+    return cls(url, params=params)
+
+
+@Environment.register_environment
+class PythonSDKEnvironment(Environment):
+
```

Comment:
add `urn = None` here and raise a `ValueError` if `self.urn is None` in `to_runner_api_parameter`

Reply:
forget the first part, it's already on `Environment`.